### PR TITLE
functions: new way to describe utask executors using other executor, with yaml

### PR DESCRIPTION
**Pull request by ybriffa**
* **What kind of change does this PR introduce?**
It introduces a new feature: functions.

* **What is the current behavior?** (You can also link to an open issue here)
There is no functions

* **What is the new behavior (if this is a feature change)?**
Functions allows to declare a directory with function that can be re-used recursively in templates to hold piece of config and abstract behavior

* **Does this PR introduce a breaking change?** (What changes might users need to make in their application due to this PR?)
no

* **Other information**:

### CODE REVIEW

**Comment by ovh-cds** [comment]
CDS Report test#609.0 ✘  
* Stage 1
  * Test ✘

File: engine/step/step.go [hunk 1]
```diff
@@ -263,14 +289,14 @@ func Run(st *Step, baseConfig map[string]json.RawMessage, values *values.Values,
 			st.State = StateToRetry
 		default:
 			st.Output, st.Metadata, st.Tags, err = runner.Exec(st.Name, baseCfgRaw, config, ctx)
-			if baseOutput != nil {
-				if st.Output != nil {
+			if len(baseOutputs) > 0 && st.Output != nil {
```
Comment (rbeuque74) [inline engine/step/step.go:292]: We could have some `baseOutputs` without having `output` from the actual runner. In that case, we should apply `baseOutputs` anyway

File: engine/step/step.go [hunk 2]
```diff
@@ -169,62 +170,87 @@ func Run(st *Step, baseConfig map[string]json.RawMessage, values *values.Values,
...
-		st.State = StateFatalError
-		st.Error = err.Error()
-		go noopStep(st, stepChan)
-		return
+		// Check if we have a function as runner or not. If not, we do not need to go further in the resolution
+		functionRunner, ok := runner.(*functions.Function)
+		if !ok {
+			break
+		}
+		var functionIntput map[string]interface{}
```
Comment (rbeuque74) [inline engine/step/step.go:244]:
```suggestion
		var functionInput map[string]interface{}
```

Comment (rbeuque74) [reply engine/step/step.go:239]: Still present

File: docker-compose.yaml [hunk 3]
```diff
@@ -16,6 +16,7 @@ services:
     volumes:
       - "./templates:/app/templates:ro"
       - "./scripts:/app/scripts:ro"
+      - "./services:/app/services:ro"
```
Comment (rbeuque74) [inline docker-compose.yaml:19]:
```suggestion
      - "./functions:/app/functions:ro"
```

**Pull request merged**